Stop throwing exception when writing 0 length buffers

TanayParikh opened the pull request:
Fixes: https://github.com/dotnet/aspnetcore/issues/31299

TanayParikh committed:
Stop throwing exception when writing 0 length buffers

Fixes: https://github.com/dotnet/aspnetcore/issues/31299

TanayParikh commented on src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs:
```diff
@@ -109,6 +109,11 @@ public override void Write(char[] buffer, int index, int count)
                 throw new ArgumentNullException(nameof(buffer));
             }
+            
+            if (index == 0 && count == 0)
```
Not sure of the exact approach / level of complexity we'd like to add here. We could do something a bit fancier like:

```suggestion
            if (index >= 0 && index < buffer.Length && count == 0)
```

pranavkm replied:
We could simply copy the checks from here: https://github.com/dotnet/runtime/blob/7938f9d9fadc3952d792329be602668bef124f23/src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs#L159

TanayParikh replied:
Thanks, updated & added a test.

TanayParikh committed:
Leverage runtime checks & add test

https://github.com/dotnet/runtime/blob/7938f9d9fadc3952d792329be602668bef124f23/src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs#L159

pranavkm commented on src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs:
```diff
@@ -110,16 +110,21 @@ public override void Write(char[] buffer, int index, int count)
                 throw new ArgumentOutOfRangeException(nameof(count));
             }
 
+            if (buffer.Length - index < count)
```
Hmm, isn’t this basically the same check as we had before?

GrabYourPitchforks commented:
See also my comment at https://github.com/dotnet/aspnetcore/issues/31299#issuecomment-808766264. You may be leaving some performance on the floor given the current method implementations.

TanayParikh commented:
> See also my comment at [#31299 (comment)](https://github.com/dotnet/aspnetcore/issues/31299#issuecomment-808766264). You may be leaving some performance on the floor given the current method implementations.

Thanks @GrabYourPitchforks, I created https://github.com/dotnet/aspnetcore/issues/32284 to track that separately.
